feat(types): add json path type inference

avallete opened the pull request:
## What kind of change does this PR introduce?

Allow users to override Json type with custom types, and take into account the "json accessor" to infer the proper resulting type.

## What is the current behavior?

Given a table defined like so:

```
type CustomJsonType = {
  foo: string
  bar: {
    baz: number
  }
  en: 'ONE' | 'TWO' | 'THREE'
}
table: {
 data: CustomJsonType | null
}
```

The following behavior is observed:

```ts
select(`data`) -> {data: CustomJsonType | null}
select(`data->bar`) -> {bar: Json}
select(`data->>bar`) -> {bar: string}
select(`data->>en`) -> {bar: string}
select(`data->bar->baz`) -> {baz: Json}
```

## What is the new behavior?

```ts
select(`data`) -> {data: CustomJsonType | null}
select(`data->bar`) -> { bar:  { baz: number } }
// Preserve the postgrest casting behaviour due to ->> accessor
select(`data->>bar`) -> {bar: string}
// Preserve the union of string
select(`data->>en`) -> {bar:  'ONE' | 'TWO' | 'THREE'}
// Preserve the accessed ressource type
select(`data->bar->baz`) -> {baz: number}
// Fallback on the default behavior for non-matching properties
select(`data->bar->nope`) -> {nope: Json}
select(`data->bar->>nope`) -> {nope: string}
```

## Additional context

Add any other context or screenshots.

avallete commented on src/select-query-parser/parser.ts:
```diff
@@ -220,13 +221,24 @@ type ParseNonEmbeddedResourceField<Input extends string> = ParseIdentifier<Input
+                name: Name
+                alias: PropertyName
+                castType: PropertyType
+                jsonPath: JsonPathToAccessor<
```
We use this to format and save the value "path accessor" information from the parser to the result inference. That way, the parser can still be agnostic to the database types, only caring about string parsing.

avallete commented on src/select-query-parser/utils.ts:
```diff
@@ -544,3 +544,35 @@ export type FindFieldMatchingRelationships<
     }
   : SelectQueryError<'Failed to find matching relation via name'>
+
+export type JsonPathToAccessor<Path extends string> = Path extends `${infer P1}->${infer P2}`
```
Turn something like `data->some->>property` into `some.property`

avallete commented on src/select-query-parser/utils.ts:
```diff
@@ -544,3 +544,35 @@ export type FindFieldMatchingRelationships<
+  ? JsonPathToAccessor<P1>
+  : Path
+
+export type JsonPathToType<T, Path extends string> = Path extends ''
```
Parse some path like `some.property` and try to match it with the type T provided.

avallete commented on test/index.test-d.ts:
```diff
@@ -52,87 +52,93 @@ const postgrest = new PostgrestClient<Database>(REST_URL)
-    const { data, error } = await postgrest.from('users').select('status').eq('status', 'ONLINE')
-    if (error) {
-      throw new Error(error.message)
+    const result = await postgrest.from('users').select('status').eq('status', 'ONLINE')
```
Here for some reason (maybe TS upgrade) using this syntax:

```
const { data, error } = await ...

if (error) {
 throw XXX
}
// here data was still of type data[] | null
// because typescript seemed to fail to understand the branching
```

Changing to this syntax properly narrow down `result.data` as non nullable once we checked the error.

kamilogorek commented:
That's already a black-magic level, but validated behavior on our live project and everything works as expected.

kamilogorek commented:
Gentle ping @soedirgo

soedirgo commented:
Seems to not work on embedded tables:

```
    const result = await postgrest
      .from('messages')
      .select('users(data->bar)')
```

(I updated the types for public.users.data with `CustomUserDataType`)

But that can be handled in a separate PR 👍

avallete commented:
> Seems to not work on embedded tables:
> 
> ```
>     const result = await postgrest
>       .from('messages')
>       .select('users(data->bar)')
> ```
> 
> (I updated the types for public.users.data with `CustomUserDataType`)
> 
> But that can be handled in a separate PR 👍

Good catch, it was an issue with the parser and the closing parenthesis, I fixed it here: https://github.com/supabase/postgrest-js/pull/592/commits/a9842b27240948a1dbe3fb602af597f13584a951

And added appropriate tests for coverage 👍
